### DISCUSSION TITLE
Set the Icon of Callout Block

### DISCUSSION THREAD

**Discussion by RoyiAvital** [quarto-dev/quarto-cli#3563] [Q&A] [answered]
It would be great to have control over the callout block.

If not, at least add a callout with question mark to allow _questions_ in the context of presentation for teaching.

**Comment by mcanouil** [comment, marked as answer]
Set icon to false and use any of the Quarto shortcodes extensions to add a new icon using level 2 header inside the callout.

```md
::: {.callout-note icon=false}

## {{< iconify fa6-brands fa-circle-question >}} Question

Using callouts is an effective way to highlight content that your reader give special consideration or attention.

:::

**Reply by cscheid** [reply]
Hey, that's a nice trick! You just saved me the couple of hours of implementing this on Callouts :)

**Reply by RoyiAvital** [reply]
@mcanouil , Could you explain the trick here?
It seems it might be useful in other cases as well.

**Reply by RoyiAvital** [reply]
OK, I tried it, this is what I get:

![image](https://user-images.githubusercontent.com/7347975/205574553-dbcdcaa4-2da9-4ccf-9e12-8394272ba9bd.png)

I don't see any icon, any idea?

**Reply by mcanouil** [reply]
Did you install the Quarto extension of your choice to get the icon from?

See for example <https://github.com/mcanouil/quarto-iconify>.

See documentation for callout structure/syntax <https://quarto.org/docs/authoring/callouts.html#markdown-syntax>.

**Reply by RoyiAvital** [reply]
Yes I did, it seems the icon you specified doesn't work. Others do. Thank You!

@cscheid , I think it is better have a built in option which is aliened with the theme.
For instance, the current set of icons match in color to the callout border.

**Reply by RoyiAvital** [reply]
@mcanouil , Is there a way to control the color? To match the theme?

**Reply by RoyiAvital** [reply]
OK, I managed to get something with `[{{< iconify fluent chat-bubbles-question-24-regular >}}]{style="color: #02b875;"}`.

**Reply by mcanouil** [reply]
This is the way to do it in Quarto.

**Reply by mcanouil** [reply]
To note, this is the same trick as in #2402

@RoyiAvital Maybe change the category to Q&A and mark the answer if you are happy with the solution proposed here.

**Reply by RoyiAvital** [reply]
@mcanouil , I meant something built in. For instance, it will automatically set the correct color for the icons, etc...

By the way, Do you know why it doesn't find `{{< iconify fa6-brands fa-circle-question >}}`?

If you want, I can open a question which this trick can be an answer to.

**Reply by mcanouil** [reply]
No need to create another discussion.

A built-in solution would be exactly like the navbar icons, thus limited to what is embedded in Quarto.
If you want other icons, then the only way is shortcode.
You could use (S)CSS to retrieve the colour (not sure if it is stored in a variable though).

**Reply by mcanouil** [reply]
I did not check the "free" part and the availability within the Iconify framework of the Font Awesome 6 icon set.
Check Iconify website to see all the icons set (not only Font Awesome).

**Reply by RoyiAvital** [reply]
> A built-in solution would be exactly like the navbar icons, thus limited to what is embedded in Quarto.
If you want other icons, then the only way is shortcode.
You could use (S)CSS to retrieve the colour (not sure if it is stored in a variable though).

I think differently.
You could also make the same policy about the whole callout, doing just one and let the user the customization of all other variants.
I guess the point of Quarto is the right balance between abstraction and simplicity.
The end user of Quarto isn't a CSS / HTML / JS wizard, it is a technical writer.

On the 1st level I think a callout specific for question will be a great and useful addition.
I also think that the `icon` field could get a value of icon instead of just `false` and `true`.
This is a good level of abstraction while keeping things simple.

**Reply by mcanouil** [reply]
You did not get my point.
The only icons that would be internally available to "icon" in callout would be the same that are natively available in navbar icon parameter.

For everything else, you will have no other choice beside using a shortcode thus in a text field option, i.e., the trick I highlighted here.

Making a different implementation without changing the navbar icons behaviour will create discrepancies and will likely mislead users.

(I agree that icon could probably get icon value as it is done in navbar icon)

**Reply by turcotte** [reply]
Do you know why the Font Awesome right quote does not appear in the first example below, but it does in the second? My main concern is how to make the right quote display correctly in the first example.

```
::: {.callout-note icon=false}
## {{< fa quote-right >}}

Some text.

:::
```

```
::: {.callout-note icon=false}
## {{< fa quote-right >}} Citation

Some text.

:::
```

**Reply by cscheid** [reply]
That's happening because Quarto attempts to detect an empty title to replace it with the placeholder title ("Note" in your case). Because the icon shortcode doesn't produce any text, Quarto thinks it's an empty title. That's a bug we need to fix.

In the meantime, though, you can use the following:

````
::: {.callout-note icon=false}
## {{< fa quote-right >}} &nbsp;

Some text.

:::
````

<img width="792" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/99bf79a7-1cb8-4d8c-b3ca-9a9a98a8fc05">

**Reply by cscheid** [reply]
Tracking the bug here https://github.com/quarto-dev/quarto-cli/issues/9701

**Reply by turcotte** [reply]
Thank you for the prompt response and solution!

**Reply by jimjam-slam** [reply]
If you don't want to install an extension, you could also use something like this to use a [Bootstrap icon](https://icons.getbootstrap.com/#icons) (which is the set used for navbars):

```
`<i class="bi bi-exclamation-diamond-fill"></i>`{=html}
```

I believe Bootstrap icons are loaded with Quarto so long as you don't use `theme: none` (although if you do, you can still manually load the icon CSS back with `format.html.css`).

**Comment by nichtich** [comment]
It's a bit tricky but you can replace the icon with an SVG (e.g. from [Boostrap Icons](https://icons.getbootstrap.com/)) via CSS, e.g.

~~~css
div.callout-note:not(.callout-titled) .callout-icon::before {
  background-image: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="%232373cc" class="bi bi-arrow-right-circle" viewBox="0 0 16 16"><path fill-rule="evenodd" d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8m15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0M4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5z"/></svg>');
}
~~~

A simple configuration key such as `callout-note-icon: arrow-right-circle` would be much more convenient.

**Reply by mcanouil** [reply]
The icon option in (custom) call-outs already controls Bootstrap Icons.

- https://quarto.org/docs/prerelease/1.3/custom-ast-nodes/callout.html

**Reply by nichtich** [reply]
Thanks! This seems pretty new and requires a custom lua filter. A minimal filter code to change the icon of a selected callout type would fully answer this question.

**Reply by mcanouil** [reply]
See extensions such as https://github.com/coatless-quarto/custom-callout

**Reply by dmi3kno** [reply]
> See extensions such as https://github.com/coatless-quarto/custom-callout

Is there a chance that custom callout could be integrated into Quarto core functionality? It's such a useful feature and I'd like to be able to define custom callouts in my teaching materials and just use them without worrying about the extension files. This is especially important when sharing Quarto documents with students.

Or perhaps it could be part of _brand.yml?

**Reply by mcanouil** [reply]
Regarding callout support in **brand.yml** that's currently a limitation regardless of the ability to fully customise callouts.
See the "Limitations" callout: https://quarto.org/docs/authoring/brand.html#overview

Custom callout API is already available for anyone to use (although, it's a bit hidden in the documentation currently, i believe there is opened issues abut this): 
- https://quarto.org/docs/prerelease/1.3/ast.html#example-callouts

This is how the above-mentioned extension is built.

Finally, about a possible integration of "custom-callout" extension within Quarto CLI itself, that would be mostly up to @cscheid, and of course @coatless.

**Reply by coatless** [reply]
No issues on my end w/ inclusion. At some point we were suppose to chat about custom AST referencing, but I think time got away from us.
